change(state): Deduplicate note commitment trees stored in the finalized state

upbqdn opened the pull request:
## Motivation

Close #4784.

Previous PRs that address parts of #4784:

- #7239.
- #7266.

This is a final PR that prunes the trees stored in the database.

## Solution

- Add support for deleting Sapling and Orchard trees in `DiskWriteBatch`.
- Prune duplicate Sapling and Orchard trees starting from the genesis block to the initial tip.
- Zebra doesn't store Sprout trees by height, it keeps only a single tree for the tip, so this PR doesn't prune Sprout trees.
- I kept the database version at 25.1.0, which was introduced in #7266. The reason is that both this PR and #7266 leave the database in the same format&mdash;this PR affects already stored trees, and #7266 affects incoming trees.

### Reviewer Checklist

  - [ ] Will the PR name make sense to users?
    - [ ] Does it need extra CHANGELOG info? (new features, breaking changes, large changes)
  - [ ] Are the PR labels correct?
  - [ ] Does the code do what the ticket and PR says?
    - [ ] Does it change concurrent code, unsafe code, or consensus rules?
  - [ ] How do you know it works? Does it have tests?

upbqdn committed:
Add support for deleting the trees

upbqdn committed:
Prune the trees

arya2 reviewed (dismissed):
This looks perfect - thank you!

arya2 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -338,4 +338,22 @@ impl DiskWriteBatch {
+            .cf_handle("orchard_note_commitment_tree")
+            .unwrap();
+        self.zs_delete(&orchard_tree_cf, height);
+    }
```
Optional: Add methods for getting the cf handles.

```suggestion
    /// Deletes the Sapling note commitment tree at the given [`Height`].
    pub fn delete_sapling_tree(&mut self, zebra_db: &ZebraDb, height: &Height) {
        self.zs_delete(&zebra_db.sapling_tree_cf(), height);
    }

    /// Deletes the Orchard note commitment tree at the given [`Height`].
    pub fn delete_orchard_tree(&mut self, zebra_db: &ZebraDb, height: &Height) {
        self.zs_delete(&zebra_db.orchard_tree_cf(), height);
    }
```

upbqdn replied:
Oh, nice. I think we can do that in other places as well.

arya2 replied:
> Oh, nice. I think we can do that in other places as well.

This came out of mild disappointment seeing the column names defined in several places - I think we should use constants there too. 

Let's merge this PR and create a ticket for that cleanup?

gustavovalverde commented:
@mergify refresh

arya2 commented:
@Mergifyio update

upbqdn committed:
Remove `Network` from `DiskWriteBatch`

Removing the `Network` from `DiskWriteBatch` makes it easy to
instantiate `DiskWriteBatch`es in `ZebraDb` that remove individual note
commitment trees.

The `Network` from `DiskWriteBatch` was used only for transparent
addresses, so the refactor isn't large. After removing it from
`DiskWriteBatch`, I passed it as a function argument instead. However,
we should simplify the parameter lists because at least two functions
have more than seven parameters now.

upbqdn committed:
Support individual tree removal in `ZebraDb`

upbqdn committed:
Refactor the tree removal task

upbqdn committed:
Prune old comments

upbqdn committed:
Remove redundant code

upbqdn commented:
I pushed a new version of the tree pruning task. The previous task had two issues:

1. When the task got canceled, and the upgrade height hadn't reached the finalized tip, the database was still marked as upgraded, and Zebra wouldn't launch the upgrade anymore, leaving the rest of the database untouched.
2. The upgrade task was using `DiskWriteBatch`es directly, which works only in tests since we don't impl [Deref](https://doc.rust-lang.org/beta/core/ops/deref/trait.Deref.html)<Target = [DiskDb](https://doc-internal.zebra.zfnd.org/zebra_state/service/finalized_state/disk_db/struct.DiskDb.html)> for `ZebraDb` in production, so I moved the `DiskWriteBatch` instantiation to `ZebraDb`. I removed `Network` from `DiskWriteBatch` to do that.

The current task seems to be working, but it's very slow. Here's the progress by 10_000 heights:

```log
2023-08-10T22:05:13 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 10000
2023-08-10T22:09:26 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 20000
2023-08-10T22:16:42 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 30000
2023-08-10T22:27:03 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 40000
2023-08-10T22:40:30 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 50000
2023-08-10T22:57:12 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 60000
2023-08-10T23:17:32 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 70000
2023-08-10T23:41:36 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 80000
2023-08-11T00:08:32 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 90000
2023-08-11T00:38:12 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 100000
2023-08-11T01:10:45 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 110000
2023-08-11T01:45:54 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 120000
2023-08-11T02:23:46 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 130000
2023-08-11T03:04:15 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 140000
2023-08-11T03:47:28 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 150000
2023-08-11T04:33:31 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 160000
2023-08-11T05:22:14 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 170000
2023-08-11T06:13:25 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 180000
2023-08-11T07:07:06 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 190000
2023-08-11T08:03:14 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 200000
2023-08-11T09:01:50 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 210000
2023-08-11T10:03:08 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 220000
2023-08-11T11:07:05 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 230000
2023-08-11T12:13:44 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 240000
2023-08-11T13:23:02 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 250000
2023-08-11T14:35:07 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 260000
2023-08-11T15:49:41 zebra_state::service::finalized_state::disk_format::upgrade: The upgrade is at height 270000
```

Notice that the upgrade task went through heights 0-10_000 in ~ 4 minutes, but it took more than an hour to go through 260_000-270_000. The task was getting slower as it progressed.

I'm currently testing another version of the upgrade task with batched removals.

upbqdn committed:
Batch the removals

arya2 committed:
delete ranges before relevant network upgrades

arya2 committed:
moves prev_tree inits

arya2 committed:
add iterator methods for reading note commitment trees

arya2 committed:
Sets up skeleton of sapling pipeline

arya2 committed:
Replaces .filter with .take_while

Fills in pipeline

Reuses zs_range_iter instead of repeating that code

Updates logic to stop at initial tip height

arya2 committed:
uses std threads

arya2 committed:
delete_range excludes end key

arya2 committed:
fixes off by one bugs

arya2 committed:
Log warning when a send fails

arya2 committed:
Removes progress logs

arya2 committed:
Log join errors instead of panicking

arya2 commented:
@Mergifyio update

upbqdn commented:
I'm doing more manual testing, so I temporarily converted this PR to a draft.

upbqdn committed:
Revert: Make the `db` field of `ZebraDb` private

upbqdn committed:
Move `delete_range_sapling_tree`

upbqdn committed:
Remove a redundant `else if` branch

Rationale:

The condition `n == 1` for the removed branch is true for a subset of
values of `n` in the preceding condition `n >= 1`.

upbqdn committed:
Use more specific error messages

upbqdn committed:
Revert: Remove redundant methods for tree removal

upbqdn commented on zebra-state/src/service/finalized_state/zebra_db/block.rs:
```diff
@@ -388,6 +389,11 @@ impl ZebraDb {
         Ok(finalized.verified.hash)
     }
+
+    /// Writes the given batch to the database.
+    pub(crate) fn write_batch(&self, batch: DiskWriteBatch) -> Result<(), rocksdb::Error> {
+        self.db.write(batch)
+    }
```
We're exposing direct write access to the database via arbitrary batches here. It would be better if we prevented that, but I can't think of a better solution.

arya2 replied:
Would it help to move methods here so this type defines all of the possible db operations that Zebra may make?

upbqdn replied:
I don't know.

arya2 replied:
Optional: open an issue with the problem? It may be worth fixing ahead of more format upgrades after the subtree indexes, or before stabilizing the zebra-state crate.

arya2 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +263,165 @@ impl DbFormatChange {
+
+                    if num_entries.map_or(false, |n| n >= 1) {
+                        let mut batch: DiskWriteBatch = DiskWriteBatch::new();
+                        batch.delete_range_orchard_tree(&db, &delete_from, &height);
```
Nitpick: it may be slightly faster to use `zs_delete` when the range would only delete a single key.

upbqdn replied:
OK, I'll bring it back.

arya2 replied:
Wait but with that said, great cleanup - please bring it back without the duplicate `batch` definitions and `write_batch` calls.

upbqdn replied:
Done in [235e43b](https://github.com/ZcashFoundation/zebra/pull/7312/commits/235e43bf8f98a7442cc885147a42c03102d0a07b).

arya2 committed:
Suggestions for Deduplicate note commitment trees stored in the finalized state (#7330)

upbqdn committed:
Add TODOs to some `Height` methods

upbqdn committed:
Add methods for deleting individual trees

upbqdn committed:
Refactor the tasks for deleting trees

upbqdn commented:
> The current task seems to be working, but it's very slow.

The migration is now blazing fast.

teor2345 commented on zebra-state/src/service/finalized_state/disk_db.rs:
```diff
@@ -457,7 +465,7 @@ impl DiskDb {
             .map(|(key, value)| (key.to_vec(), value))
             // Handle Excluded start and the end bound
-            .filter(move |(key, _value)| range.contains(key))
+            .take_while(move |(key, _value)| range.contains(key))
```
Low priority bug:

This will terminate iteration too early if there's an excluded start bound.

I don't think this causes any bugs in existing code, because excluded start bounds don't have a standard Rust iteration syntax. (So you'd have to pass `(Excluded(start), _)` to trigger the bug.)

Instead, we probably want something like:
```rust
// skip excluded start bound (and empty ranges), the start_mode already skips keys before the start bound
.skip_while(move |(key, _value)| !range.contains(key))
// take until the excluded end bound is reached, or we're after the included end bound
.take_while(move |(key, _value)| range.contains(key))
```

arya2 replied:
Won't the take_while predicate always return true if it skips elements that would return false?

teor2345 replied:
I'm pretty sure the code works as documented. But it's a good question, because there aren't any tests. We might want to add them as part of PR #7365 or Marek's upcoming fix PR. (If the tests in those PRs don't cover this code already.)

### Detailed Background

Have you read the `RangeBounds`, `skip_while()`, and `take_while()` docs?

A `RangeBounds` identifies a continuous sequence of ordered keys, but it can exclude some keys at the start and end of the column family:
https://doc.rust-lang.org/std/ops/trait.RangeBounds.html

`skip_while()` runs the predicate on each item it gets, skipping items until it fails, then yields the rest of the iterator (without running the predicate):
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.skip_while

`take_while()` runs the predicate on each item it gets, yielding items until it fails, then ignores the rest of the iterator:
https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.take_while

<details>

<summary>detailed explanation of this implementation</summary>

There are 3 kinds of key exclusion that happen in this code:
At the database level:
- `start_mode` can skip some keys at the start, but always includes the start bound, even if it should be excluded by the range
At the iterator level:
- `skip_while()` skips the start bound if it should be excluded, then includes everything from the first key that is contained in the range (ignoring the predicate). It should only skip zero or one key (because start_mode skips most of them). Then it includes all the remaining keys, regardless of the end of the range.
- `take_while()` includes everything until the end bound is reached, then stops the iterator (correctly implementing inclusion or exclusion). It can include any number of keys, then stop before any number of keys. 

</details>

arya2 replied:
Ah, for some reason I was imagining it doing the same thing as filter. I got it now.

upbqdn replied:
Addressed in [0b3a1b7](https://github.com/ZcashFoundation/zebra/pull/7379/commits/0b3a1b7ffe333b29c7ed4bfcb26c1453d4c2fa08).

teor2345 commented on zebra-state/src/service/finalized_state/disk_db.rs:
```diff
@@ -115,11 +113,17 @@ pub trait WriteDisk {
         C: rocksdb::AsColumnFamilyRef,
         K: IntoDisk + Debug;
+
+    /// Remove the given key range from rocksdb column family if it exists.
```
Doc Nit: Including `from` but excluding `to`.

upbqdn replied:
Addressed in [cd3bff6](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd3bff6668afe0917d23a4ccd5cb119660d9b70a).

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+            let should_cancel_flag = should_cancel_format_change.clone();
+            let sapling_read_task = std::thread::spawn(move || {
+                for (height, tree) in
+                    db.sapling_tree_by_height_range(sapling_height..initial_tip_height)
```
This range should be `0..=initial_tip_height`, for two reasons:
1. If the first shielded transaction is not in the activation block, then there will be duplicate empty note commitment trees for the genesis and activation blocks
2. If the initial tip height is a duplicate, it needs to be deleted (see my detailed comment on the code below)

This is a bug that requires us to re-run the upgrade task after fixing it, by assigning it a new state version.

teor2345 replied:
We might also want to check for a similar missed duplicate bug in the first new block that is committed to the finalized state after Zebra is started.

upbqdn replied:
Addressed in [cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf).

> We might also want to check for a similar missed duplicate bug in the first new block that is committed to the finalized state after Zebra is started.

I don't think this can happen because duplicate trees aren't committed to the database anymore: https://github.com/zcashfoundation/zebra/blob/ead854b587d9a64ef1ecd5c4971b3a4736c04a91/zebra-state/src/service/finalized_state/zebra_db/shielded.rs#L339-L356.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+            let db = upgrade_db.clone();
+            let sapling_delete_task = std::thread::spawn(move || {
+                let mut delete_from = sapling_height.next();
+                while let Ok(delete_to) = unique_sapling_tree_height_rx.recv() {
```
If the final `delete_to` is not equal to the initial tip height, there needs to be a delete range from the final `delete_from`, up to and including the initial tip height. (So to `initial_tip_height + 1`.)

This is a bug that requires us to re-run the upgrade task after fixing it, by assigning it a new state version.

upbqdn replied:
Addressed in https://github.com/zcashfoundation/zebra/blob/ead854b587d9a64ef1ecd5c4971b3a4736c04a91/zebra-state/src/service/finalized_state/zebra_db/shielded.rs#L339-L356.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+                orchard_delete_task,
+            ] {
+                if let Err(error) = task.join() {
+                    warn!(?error, "unexpected join error")
```
Low priority bug:

If a thread errors that means that its code has panicked. So the error handler should resume the panic unwind, just logging a warning can let code in threads panic undetected.

In particular, the database would be marked as upgraded even if a panic terminated the upgrade early, which is a bug.

Instead, let's just simplify things by removing the extra threads?

If there are actually any panics in these threads, then we might need to re-run the format upgrade after fixing the panics and this panic handler.

teor2345 replied:
Optional:

#### Related Logging Bug

Threads need to be explicitly set up with the logging spans, otherwise their logs can be missing important diagnostic info.

#### Related Performance Question

Are these threads actually more efficient than running all the code in the upgrade thread? Does the difference matter?

Splitting code into separate threads can actually slow it down due to lock contention, context switching overhead, and missed compiler optimisations. (Code that's split across threads can't be optimised using assumptions from code in another thread.)

upbqdn replied:
[cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf) removes the threads.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
-            // Keep upgrading until the initial database has been upgraded,
-            // or this task is cancelled by a shutdown.
-            while upgrade_height <= initial_tip_height
-                && matches!(cancel_receiver.try_recv(), Err(mpsc::TryRecvError::Empty))
```
Low priority bug:

The channel check cancels the task when the sender is dropped, but the AtomicBool does not.

That might make Zebra hang on shutdown waiting for a format upgrade task to finish.

One way to fix this is to revert to the channel check (we can use a wrapper, trait, or function to simplify it if needed). Another is to set the bool on drop, but in general I try not to use atomic because I don't understand how atomic orderings work and how to avoid concurrency bugs when using them. So it's tricky for me to review this code.

upbqdn replied:
[cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf) uses `mpsc`.

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -112,8 +112,7 @@ impl ZebraDb {
 
-    /// Returns the Sapling note commitment tree of the finalized tip
-    /// or the empty tree if the state is empty.
+    /// Returns the Sapling note commitment trees starting from the given block height up to the chain tip
```
Doc Nit: I think this comment update is a copy-paste error, there's no height passed to this method.

upbqdn replied:
Addressed in [c831a74](https://github.com/ZcashFoundation/zebra/pull/7379/commits/c831a74d9481cf830309c83a8e06e6578796628d).

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -124,6 +123,33 @@ impl ZebraDb {
             .expect("Sapling note commitment tree must exist if there is a finalized tip")
     }
 
+    /// Returns the Orchard note commitment trees starting from the given block height up to the chain tip
```
Doc Nit: The range can have an end bound, so this doc should be "in the supplied range", and the same for the sapling method.

upbqdn replied:
Addressed in [9dba249](https://github.com/ZcashFoundation/zebra/pull/7379/commits/9dba2492d920ef9e5728b5543427aa2b63ba0079).

teor2345 commented on zebra-state/src/service/finalized_state/zebra_db/shielded.rs:
```diff
@@ -338,4 +357,40 @@ impl DiskWriteBatch {
+    }
+
+    /// Deletes the range of Sapling note commitment trees at the given [`Height`]s. Doesn't delete the upper bound.
+    pub fn delete_range_sapling_tree(&mut self, zebra_db: &ZebraDb, from: &Height, to: &Height) {
```
Optional:

It seems like this API caused a state format bug, maybe it would be better to write it in terms of a Rust `RangeBounds`?

(Like `sapling_tree_by_height_range`.)

This would be best implemented by changing `zs_delete_range` to take a Rust `RangeBounds`.

teor2345 replied:
Edit: `RangeBounds` is the generic trait, `Range` is a particular type of range that implements it.

upbqdn replied:
I tried implementing this but ended up not including it in PR #7379 because it would make the current pruning a bit more complicated than it is.

teor2345 replied:
No worries! Happy to open a ticket or add a TODO or just do nothing until the refactor is needed. Whatever you think is best.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+        // Start of a database upgrade task.
+
+        let version_for_pruning_trees =
+            Version::parse("25.1.0").expect("Hardcoded version string should be valid.");
```
Since PR #7266 was merged into the main branch before this PR, some cached states might have already been marked as "upgraded" to version `25.1.0`, without actually doing the upgrade in this PR's code.

That means that CI could be failing on the note commitment subtrees PRs, because the data format is not as expected.

Let's upgrade to patch version 25.1.1 in the fix PR?

(This is a bug that requires a re-run of the upgrade task, but only in Zebra's CI.)

upbqdn replied:
Addressed in [5c299cc](https://github.com/ZcashFoundation/zebra/pull/7379/commits/5c299cc6355cd435ee14a2ace778400849239f0d).

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+            info!(?newer_running_version, "Database has been upgraded to:");
+            Self::mark_as_upgraded_to(&version_for_pruning_trees, &config, network);
         }
 
```
Since there have been multiple bugs in this code which the tests didn't catch, let's do a runtime test here that the format is actually as expected?

If we've actually deleted all the duplicates, this should be quick.

I'll write this test in a separate PR.

teor2345 replied:
This test is now in PR #7365.

teor2345 commented on zebra-state/src/service/finalized_state/disk_format/upgrade.rs:
```diff
@@ -261,46 +269,187 @@ impl DbFormatChange {
+                let height = Height(1);
+                let mut batch = DiskWriteBatch::new();
+
+                batch.delete_range_sapling_tree(&upgrade_db, &height, &sapling_height);
```
This code is a nice optimisation, but it can leave a duplicate tree in the network upgrade activation block, if that block contains no shielded transactions. (Which is the typical case, because wallets often wait for activation before generating those transactions.)

I'd suggest deleting this code to reduce code complexity, even if it is slightly slower.

This is a bug that requires us to re-run the upgrade task after fixing it, by assigning it a new state version.

upbqdn replied:
Addressed in [cd4be3e](https://github.com/ZcashFoundation/zebra/pull/7379/commits/cd4be3e2f6829994c7b7c2989f5b39fd06141ecf).

teor2345 reviewed (commented):
I think there are still some duplicate note commitment trees after this state upgrade. See my tests in PR #7365.

I've tried to mark each review comment as a breaking format change, low priority related issue, or nitpick.

teor2345 commented:
When I run this test on my fully synced `v25.0` state, I get these logs:
```
2023-08-22T05:08:00.359950Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate sapling trees after running de-duplicate tree upgrade height=Height(419200) prev_height=Height(0) tree_root=Root("fbc2f4300c01f0b7820d00e3347c8da4ee614674376cbc45359daa54f9b5493e")

2023-08-22T05:08:02.610711Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(1687104) prev_height=Height(0) tree_root=Root("ae2935f1dfd8a24aed7c70df7de3a668eb7a49b1319880dde2bbd9031ae5d82f")
2023-08-22T05:08:03.271360Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199893) prev_height=Height(2199892) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271380Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199894) prev_height=Height(2199893) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
```

<details>
<summary>8 duplicate orchard trees from 2199895-2199902</summary>
```
2023-08-22T05:08:03.271385Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199895) prev_height=Height(2199894) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271388Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199896) prev_height=Height(2199895) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271392Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199897) prev_height=Height(2199896) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271395Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199898) prev_height=Height(2199897) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271399Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199899) prev_height=Height(2199898) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271402Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199900) prev_height=Height(2199899) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271406Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199901) prev_height=Height(2199900) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271412Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199902) prev_height=Height(2199901) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
```
</details>

```
2023-08-22T05:08:03.271416Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199903) prev_height=Height(2199902) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")
2023-08-22T05:08:03.271420Z ERROR zebra_state::service::finalized_state::disk_format::upgrade: found duplicate orchard trees after running de-duplicate tree upgrade height=Height(2199904) prev_height=Height(2199903) tree_root=Root("ffb1812f72a3888d8ccc307ca4701078daa6df664e5766ccc3ead1764ed29727")

The application panicked (crashed).
Message:  found duplicate sapling or orchard trees after running de-duplicate tree upgrade
Location: zebra-state/src/service/finalized_state/disk_format/upgrade.rs:489

Metadata:
version: 1.1.0+41.g54ce3ad
Zcash network: Mainnet
running state version: 25.1.0
initial disk state version: 25.0.2
```

<details>
<summary>extra metadata</summary>

```
features: default,getblocktemplate_rpcs,metrics_exporter_prometheus,prometheus,release_max_level_info 
branch: de-duplicate-test
git commit: 54ce3ad
commit timestamp: 1980-01-01T00:00:00.000000000Z
target triple: x86_64-unknown-linux-gnu
rust compiler: 1.74.0-nightly
rust release date: 2023-08-21
optimization level: 3
debug checks: true
```
</details>

Testnet has about 30 sapling duplicates, and about 10,000 orchard duplicates. (There might be an early termination or concurrency bug on orchard, or it might just be the same bugs.)

upbqdn commented:
Thanks for the review and the test. I'm going to open another PR to address the issues.

upbqdn commented:
I opened PR #7379 which addresses all comments.
